### PR TITLE
[ISSUE-756] Fix the unrelated SCs check in scheduler-extender

### LINKED ISSUE #756: PVC can be skipped in extender due to long cache updating
**Describe the bug**
PVC was skipped in extender due to SC not found
```
Feb 22 01:31:46.128 [INFO] [Extender] [gatherCapacityRequestsByProvisioner] [baremetal-injector] [d6c9bf78-62af-4acb-bd4c-c9f3f935699d] PVC csi-baremetal87vw5 skipped due to storage class is not provisioned
```
In every CI test volume is provisioned by new custom SC
```
Feb 22 01:31:46.127 [DEBU] [Extender] [gatherCapacityRequestsByProvisioner] [baremetal-injector] [d6c9bf78-62af-4acb-bd4c-c9f3f935699d] PVC csi-baremetal87vw5 status: {Phase:Pending AccessModes:[] Capacity:map[] Conditions:[]} spec: {AccessModes:[ReadWriteOnce] Selector:nil Resources:{Limits:map[] Requests:map[storage:{i:{value:104857600 scale:0} d:{Dec:<nil>} s:100Mi Format:BinarySI}]} VolumeName: StorageClassName:0xc0003a6060 VolumeMode:0xc0003a6070 DataSource:nil DataSourceRef:nil} SC: volume-8081trs2z
```
We use caching for SC in extenders, so it wasn't updated in time - https://github.com/dell/csi-baremetal/blob/62326e77875cca58a4badc2874d2bba70257ecc6/pkg/scheduler/extender/extender.go#L597

**Environment (please complete the following information):**
 - virtual

**To Reproduce**
Run CI suite

### PR DESCRIPTION
**Pull request by safronovD**
### Resolves #756

Refactor SC checking in extender:
- SC is found and provisioned by CSI Baremetal -> add volume in requests
- SC is found and provisioned by another driver -> skip volume
- SC is not found -> skip filter request (return empty list) and wait cache update

## PR checklist
- [x] Add link to the issue
- [x] Choose Project
- [x] Choose PR label
- [x] New unit tests added
- [x] Modified code has meaningful comments
- [x] All TODOs are linked with the issues
- [ ] All comments are resolved

## Testing
**CSI Baremetal PVC**
```
user@user-vm ~/g/s/g/dell> ka csi-baremetal/test/app/nginx.yaml

user@user-vm ~/g/s/g/dell> kg volume
NAME                                       SIZE        STORAGE CLASS   HEALTH   CSI_STATUS   LOCATION                               NODE
pvc-00ecbf97-c256-461a-aed4-ceec70787ffc   105906176   HDD             GOOD     PUBLISHED    91b8be6c-5f0b-4b8a-bef3-9f8fc2ecb2c2   56e4455e-2ed9-42bf-a8b9-8fcd1063008e
pvc-034a9c9e-c49f-4d4c-9353-4b8e01a7b6cd   105906176   HDD             GOOD     PUBLISHED    3a595a44-5054-45d6-883b-076d8f9c7d5f   dda09db1-0414-4525-a1f7-83c4f472e8b6
pvc-33bc57c3-84d7-4688-8cb1-3d40ddf1fddb   105906176   HDD             GOOD     PUBLISHED    7378aceb-a651-4cd4-970a-84ffde391f7a   56e4455e-2ed9-42bf-a8b9-8fcd1063008e
pvc-3efea709-7a79-438c-9c2c-cde3ccf4ee92   105906176   HDD             GOOD     PUBLISHED    4c88bfb7-c4e7-4476-a671-0b68d24cded4   dda09db1-0414-4525-a1f7-83c4f472e8b6
pvc-85c8bccc-12d3-4c7c-a7cb-df1c7ff0afc0   105906176   HDD             GOOD     PUBLISHED    948d5066-baa4-4097-8805-3a8abd243ca6   dda09db1-0414-4525-a1f7-83c4f472e8b6
pvc-b43240e7-7257-4de5-9523-bddac4cc069d   105906176   HDD             GOOD     PUBLISHED    9910d205-3957-4612-9fe1-9df05d14d0c6   56e4455e-2ed9-42bf-a8b9-8fcd1063008e
```

**CSI Baremetal generic ephemeral volumes**
```
user@user-vm ~/g/s/g/dell> ka csi-baremetal/test/app/nginx-generic-ephemeral.yaml
statefulset.apps/web created

user@user-vm ~/g/s/g/dell> kg volume
NAME                                       SIZE        STORAGE CLASS   HEALTH   CSI_STATUS   LOCATION                               NODE
pvc-4647b96b-f57b-4e58-acc8-1768552bd94e   105906176   HDD             GOOD     PUBLISHED    4c88bfb7-c4e7-4476-a671-0b68d24cded4   dda09db1-0414-4525-a1f7-83c4f472e8b6
pvc-4da08b26-c95e-4db1-88d9-19aef1c51b74   105906176   HDD             GOOD     PUBLISHED    3a595a44-5054-45d6-883b-076d8f9c7d5f   dda09db1-0414-4525-a1f7-83c4f472e8b6
pvc-9599ea6a-61b6-445f-9fb8-ac413f9bba9c   105906176   HDD             GOOD     PUBLISHED    91b8be6c-5f0b-4b8a-bef3-9f8fc2ecb2c2   56e4455e-2ed9-42bf-a8b9-8fcd1063008e
pvc-b1570d1e-5773-4855-8c1e-461d49e1de82   105906176   HDD             GOOD     PUBLISHED    948d5066-baa4-4097-8805-3a8abd243ca6   dda09db1-0414-4525-a1f7-83c4f472e8b6
pvc-b438e816-7fde-4414-b066-ce88fa0c1cd7   105906176   HDD             GOOD     PUBLISHED    7378aceb-a651-4cd4-970a-84ffde391f7a   56e4455e-2ed9-42bf-a8b9-8fcd1063008e
pvc-f797102b-a704-4ddc-9df7-e6a226d9ac80   105906176   HDD             GOOD     PUBLISHED    9910d205-3957-4612-9fe1-9df05d14d0c6   56e4455e-2ed9-42bf-a8b9-8fcd1063008e
```

**Other CSI PVC**
```
user@user-vm ~/g/s/g/dell> ka csi-baremetal/test/app/nginx.yaml 
statefulset.apps/web created

user@user-vm ~/g/s/g/dell> kg pvc
NAME         STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
data-web-0   Bound    pvc-955cba4c-6d03-4b5f-b00a-562958f57722   60Mi       RWO            standard       25s
data-web-1   Bound    pvc-0e22117e-4da5-4456-a1ef-954bed8172cb   60Mi       RWO            standard       25s
logs-web-0   Bound    pvc-709d1404-d3b9-430a-ad69-a8f5960e891c   60Mi       RWO            standard       25s
logs-web-1   Bound    pvc-66d2c9ba-efee-406e-839b-1bfcc24689fd   60Mi       RWO            standard       25s
www-web-0    Bound    pvc-db0449bb-2e29-423c-9c79-3202cd7a3f2f   60Mi       RWO            standard       25s
www-web-1    Bound    pvc-033470e1-e140-40b2-a103-7414a0841809   60Mi       RWO            standard       25s
```

**Wrong SC**
```
user@user-vm ~/g/s/g/dell> kd pvc data-web-0
Name:          data-web-0
Namespace:     default
StorageClass:  standard23
Status:        Pending
Volume:        
Labels:        app=nginx
               app.kubernetes.io/name=nginx
Annotations:   <none>
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      
Access Modes:  
VolumeMode:    Filesystem
Used By:       web-0
Events:
  Type     Reason              Age               From                         Message
  ----     ------              ----              ----                         -------
  Warning  ProvisioningFailed  4s (x2 over 14s)  persistentvolume-controller  storageclass.storage.k8s.io "standard23" not found
```

### CODE REVIEW

File: pkg/scheduler/extender/extender.go [hunk 1]
```diff
@@ -252,7 +250,20 @@ func (e *Extender) gatherCapacityRequestsByProvisioner(ctx context.Context, pod
...
 		if v.Ephemeral != nil {
 			claimSpec := v.Ephemeral.VolumeClaimTemplate.Spec
-			if storageType, ok := scs[*claimSpec.StorageClassName]; ok {
+			if claimSpec.StorageClassName == nil {
+				ll.Warningf("PVC %s skipped due to empty StorageClass", v.Ephemeral.VolumeClaimTemplate.Name)
+				continue
+			}
+
+			storageType, scType := scChecker.check(*claimSpec.StorageClassName)
+			switch scType {
```
Comment (rkiselenko) [inline pkg/scheduler/extender/extender.go:259]: Can we `DRY` those switches, the code from 259-278 and 312-332 looks like the same.

**Review by rkiselenko — CHANGES_REQUESTED** [review]

File: pkg/scheduler/extender/extender.go [hunk 2]
```diff
@@ -589,27 +609,69 @@ func nodePrioritize(nodeMapping map[string][]volcrd.Volume) (map[string]int64, i
...
-	return scNameTypeMap, nil
+
+	return result, nil
 }
 
-func generateEphemeralVolumeName(podName, volumeName string) string {
-	return podName + "-" + volumeName
+// scChecker.check return codes
+const (
+	relatedSC   = 0
```
[the remark this answers is not in the document]
Comment (safronovD) [reply pkg/scheduler/extender/extender.go]: done

File: pkg/scheduler/extender/extender.go (hunk 1 above)
Comment (safronovD) [reply pkg/scheduler/extender/extender.go:259]: We need to have different log messages and request fields for them. In my opinion, creating a common interface for just 2 calls brings more code here

Comment (rkiselenko) [reply pkg/scheduler/extender/extender.go:259]: But you can just create an anonymous function with all repeated stuff, less code is better that the same code repeated two times.

**Review by rkiselenko — CHANGES_REQUESTED** [review]

File: pkg/scheduler/extender/extender.go (hunk 1 above)
Comment (safronovD) [reply pkg/scheduler/extender/extender.go:259]: But it is used `return` and `continue` in switch cases. We should whatever return a status code and parse it in another switch in the main loop. Or what do you mean?

File: pkg/scheduler/extender/extender.go [hunk 3]
```diff
@@ -589,27 +609,69 @@ func nodePrioritize(nodeMapping map[string][]volcrd.Volume) (map[string]int64, i
 	return nrank, maxCount
 }
 
-// scNameStorageTypeMapping reads k8s storage class resources and collect map with key storage class name
-// and value .parameters.storageType for that sc, collect only sc that have provisioner e.provisioner
-func (e *Extender) scNameStorageTypeMapping(ctx context.Context) (map[string]string, error) {
-	scs := storageV1.StorageClassList{}
+func generateEphemeralVolumeName(podName, volumeName string) string {
```
Comment (rkiselenko) [inline pkg/scheduler/extender/extender.go:612]: this can be just `fmt.Sprintf(...)`

**Review by rkiselenko — CHANGES_REQUESTED** [review]

File: pkg/scheduler/extender/extender.go [hunk 4]
```diff
@@ -226,14 +226,12 @@ func (e *Extender) gatherCapacityRequestsByProvisioner(ctx context.Context, pod
 		"pod":         pod.Name,
 	})
 
-	scs, err := e.scNameStorageTypeMapping(ctx)
+	scChecker, err := e.buildSCChecker(ctx, ll)
```
[the remark this answers is not in the document]
Comment (safronovD) [reply pkg/scheduler/extender/extender.go:229]: What name do you prefer?

[the remark this answers is not in the document]
Comment (rkiselenko) [reply pkg/scheduler/extender/extender.go:229]: `scChecker` [is a name of the type](https://github.com/dell/csi-baremetal/blob/e7c2f8823e817560c933ba2f6c94c5bbf996a22d/pkg/scheduler/extender/extender.go#L618) just dont use it, use any other.

File: pkg/scheduler/extender/extender.go (hunk 1 above)
Comment (safronovD) [reply pkg/scheduler/extender/extender.go:259]: Could you please suggest another design for `scChecker` component to discuss? 

In my opinion, we return the status code from the function and just execute 2 strings - print log and make a decision about loop. In the third case `relatedSC` execution is different, so creating common function leads to more code due to interface usage and a lot of parameters.

I agree that making functions for repeatable code is a good practice and must be used as much as possible. But, from my perspective, in this case, it's not applicable.

We can add a new return code to the enum in the future and will have to change `switch`es too. So `default` section helps to simplify error locating for developer during maintaining this code. I can place `panic` here, but it's not a kubernetes way due to log corrupting

File: pkg/scheduler/extender/extender.go (hunk 3 above)
Comment (safronovD) [reply pkg/scheduler/extender/extender.go:612]: We have already discussed it later. I wished to underline here that we don't use a common name for inline volumes. Meaningful function name gives more info to the developer

File: pkg/scheduler/extender/extender.go [hunk 5]
```diff
@@ -589,27 +609,69 @@ func nodePrioritize(nodeMapping map[string][]volcrd.Volume) (map[string]int64, i
...
-	return scNameTypeMap, nil
+
+	return result, nil
 }
 
-func generateEphemeralVolumeName(podName, volumeName string) string {
-	return podName + "-" + volumeName
+// scChecker.check return codes
+const (
+	relatedSC = iota
```
Comment (mishoyama) [inline pkg/scheduler/extender/extender.go:658]: May be names like `managed` \ `unmanaged` are better. `related` \ `unrelated` sound confusing to me.

File: pkg/scheduler/extender/extender.go (hunk 1 above)
Comment (rkiselenko) [reply pkg/scheduler/extender/extender.go:259]: > Could you please suggest another design for scChecker component to discuss?

Something simpler.

> ‎creating common function leads to more code due to interface usage and a lot of parameters.
DRY is not about more code, is about less code.

> We can add a new return code to the enum in the future

[Stop Predicting the Future of Your Code](https://medium.com/swlh/stop-predicting-the-future-of-your-code-e0d403116f90)

File: pkg/scheduler/extender/extender.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (safronovD) [reply pkg/scheduler/extender/extender.go]: Oh, missed it, thank you

[the remark this answers is not in the document]
Comment (safronovD) [reply pkg/scheduler/extender/extender.go]: fixed

File: pkg/scheduler/extender/extender.go (hunk 1 above)
Comment (mishoyama) [reply pkg/scheduler/extender/extender.go:259]: May be it make sense to move switch statement to separate function.

File: pkg/scheduler/extender/extender.go (hunk 5 above)
Comment (safronovD) [reply pkg/scheduler/extender/extender.go]: Fixed

**Review by rkiselenko — APPROVED** [review]

**Review by mishoyama — APPROVED** [review]

**Review by yurkov-anton — APPROVED** [review]

**Pull request merged**